src/nspawn/nspawn-network.c

```diff
@@ -168,12 +168,31 @@ static int add_veth(
         return 0;
 }
 
+static char *shorten_ifname(char *ifname) {
+        unsigned char hash = 0xff;
+
+        assert(ifname);
+
+        if (strlen(ifname) >= IFNAMSIZ) {
+
+                /* Create a simple 8-bit hash of an interface name */
```

Comment:
hmm, with an 8 bit hash the chance of collision will grow above .5 as soon as you have ~18 containers with the common prefix. Isn't that a bit low? with 16 bit you need ~301 containers, with 24 bit ~4823. (if i did my math correctly).

I find 18 containers a bit too low. I think allowing 4K containers, i.e. 24bit hashin sounds ok'ish to me. That would mean giving up 6 characters for the hash. wouldn't that be OK?

(Alternatively, instead of encoding the hash in hex one could encode it in a larger alphabet, i.e. base64-like for example)

Reply:
Well, we only have 16 chars per interface name including the zero terminator, thus 15 usable chars. Thus, a hash needs to steal digits from those 15 chars of the interface name. Having a 24-bit checksum means we would steal 6 digits. Using a larger alphabet would definitely make sense.

I was already thinking about this. My current patch solved the problem for me, so I decided to suggest it upstream. Thanks for the review.

I'll look what I can come up with.

Maybe we could make the checksum length configurable? Defaulting to 0 and thus not breaking compatibility? Where should such an option go?

Reply:
Nah, please let's not add a config option for this. The fewer the better. Let's just pick sane behaviour and stick to it.

Reply:
@poettering What about reusing the hash made by `generate_mac()`, and just the second half of it?

Reply:
generate_mac() is just a wrapper around siphash too, and takes the machine id and interface name as input, and then makes sure the result has the right bits set to be recognizable as hashed/randomized MAC addresses. We need none of that here, besides the actual siphash.

I don't the the interface names we generate should take the machine id as input, the names should be generated the same everywhere. Hence, I don't think there#s anything to share with generate_mac(), that just calling siphash24() wouldn't already give you.